### PR TITLE
[Performance][MRV2] Remove atomic_or from the bincount prompt bitmask

### LINKED ISSUE #14158: [Contribution] [Perf][MRV2] bincount 算子性能优化
## 背景

MRV2 采样路径中，`bincount` 构建 prompt 的二进制 bitmask（`prompt_bin_mask`，标记 prompt 中出现过的 token）与输出 token 计数（`output_bin_counts`），供 `apply_penalties` 使用。Ascend 侧实现位于 `vllm_ascend/worker/v2/sample/penalties.py`。

当前实现为 Triton kernel `_bincount_kernel`，grid `(num_tokens, num_blocks)`，`BLOCK_SIZE=1024`。kernel 通过 `tl.atomic_or` 写 bitmask（token_id // 32 为字索引，token_id % 32 为位），通过 `tl.atomic_add` 累加 output_bin_counts。两个原子操作在高并发下存在竞争。

存在以下可优化点：
- `atomic_or`（bitmask）与 `atomic_add`（counts）原子操作竞争；
- `BLOCK_SIZE=1024` 固定；
- 每个 token 独立遍历 prompt，重复扫描可合并。

## 任务

优化 `bincount` 在 Ascend NPU 上的性能。
- 路径：`vllm_ascend/worker/v2/sample/penalties.py`（`_bincount_kernel` / `bincount`）
- 方向：降低原子竞争（分块规约 / 私有副本合并）、BLOCK_SIZE 自适应、减少重复扫描

### 1. 精度对比（前后）
- 优化前后 `prompt_bin_mask` 与 `output_bin_counts` 在相同输入下完全一致（bitmask 位级一致、counts 整数一致）；
- 与 vllm-core GPU 基线对齐；
- 端到端（bincount + apply_penalties 联动）输出与优化前一致。

### 2. 性能对比（前后）
- 单算子 profiling：优化前后 kernel 延迟（多组 num_tokens / prompt_len / prefill_len）；
- NPU profiler 时间线（原子操作开销）；
- 端到端：开启惩罚场景下 decode 延迟对比。

### 3. 交付件
- PR + 设计说明 + 精度对比表 + 性能数据 + 单测

## 环境约定
- vllm-ascend：最新 main
- 硬件：Ascend NPU（注明型号 + 卡数 + TP 配置）
- 关联任务池：#9079
- 验收人：@chengduxiaowu

## 重点关注

- bitmask 为位级语义，优化不得改变任一位；
- 原子规约合并后须保证 counts 正确。

## 任务周期

* 发布：2026-08-12
* 回收：2026-10-31

### PR DESCRIPTION
**Pull request by zihanlin-ai**
### What this PR does / why we need it?

Closes #14158 (MRV2 task of #9079).

Ascend `_bincount_kernel` uses `tl.atomic_or` to build `prompt_bin_mask`. On A2/A3,
this operation serializes and dominates the kernel; replacing it with a plain store in a
measurement-only kernel makes the prompt path 6.5x-33.2x faster, while removing
`atomic_add` improves the output-count path by only 1.24x-1.50x.

This PR removes `atomic_or` with two passes:

1. Write one byte per prompt vocabulary entry to a zero-initialized scratch row. Duplicate
   tokens safely write the same value.
2. Pack the byte flags into the existing int32 bitmask without atomics.

`output_bin_counts` keeps `tl.atomic_add`. The pack pass overwrites every target word, so
the old `prompt_bin_mask` row zeroing is removed. Scratch usage is
`new_penalty_requests x padded_vocab` bytes: about 0.1 MiB for one request and 9.3 MiB for
64 requests at vocab size 151936.

### Does this PR introduce _any_ user-facing change?

No. `prompt_bin_mask`, `output_bin_counts`, and penalty-adjusted logits remain numerically
equivalent to the current implementation and the vLLM-core GPU baseline.

#### Correctness tests

```bash
pytest -q tests/e2e/nightly/single_node/ops/singlecard_ops/triton/test_bincount.py
# 14 passed on A3

ruff check vllm_ascend/worker/v2/sample/penalties.py \
  tests/e2e/nightly/single_node/ops/singlecard_ops/triton/test_bincount.py
ruff format --check vllm_ascend/worker/v2/sample/penalties.py \
  tests/e2e/nightly/single_node/ops/singlecard_ops/triton/test_bincount.py
```

The test now exercises the public `bincount()` path and covers:

- scrambled, non-contiguous request mappings: `[63]`, `[63, 7, 31, 2]`, `[17]`,
  and `[40, 9]`;
- vocab sizes 151936, 151000, and 5121, including non-32-multiple tail words;
- short, multi-block, multi-request, and zero-output-token inputs;
- poisoned outputs and verification that unmapped rows remain untouched;
- `bincount -> apply_penalties` logits against a PyTorch reference.

The same inputs were run through vLLM-core `bincount` on an NVIDIA H20 with vLLM 0.27.1.
All three comparison cases matched byte for byte:

| case | vocab | request rows | prompt mask | output counts |
|---|---:|---|---|---|
| scrambled mapping | 5121 | `[63, 7, 31, 2]` | MATCH | MATCH |
| tail word | 151000 | `[63]` | MATCH | MATCH |
| full vocab, multi-block | 151936 | `[17, 40]` | MATCH | MATCH |

#### Single-operator performance

`vocab_size=151936`; times are per `bincount()` call.

| new requests | prompt | output | A2 current / PR | A2 speedup | A3 current / PR | A3 speedup |
|---:|---:|---:|---:|---:|---:|---:|
| 1 | 8 | 4 | 0.621 / 0.662 ms | 0.94x | 0.310 / 0.368 ms | 0.84x |
| 1 | 1024 | 256 | 0.867 / 0.662 ms | 1.31x | 0.572 / 0.438 ms | 1.30x |
| 1 | 65280 | 256 | 21.730 / 0.841 ms | 25.8x | 20.387 / 0.587 ms | 34.8x |
| 16 | 16384 | 256 | 86.532 / 2.461 ms | 35.2x | 82.021 / 2.223 ms | 36.9x |
| 64 | 4096 | 256 | 87.231 / 5.915 ms | 14.8x | 82.590 / 5.689 ms | 14.5x |

The extra pack launch costs at most 0.06 ms for very short prompts. Break-even is about 200
prompt tokens on both A2 and A3. `bincount()` runs only when a penalty-using request is
admitted, not on every decode step. `WORDS_PER_BLOCK=128` was retained after a sweep: 64
and 128 tie, 256 is slower, and 512 or above exceeds UB capacity.

#### NPU profiler

`torch_npu.profiler` on A3, 4 new requests x 16384 prompt tokens, 5 iterations:

| device-side total | current | this PR |
|---|---:|---:|
| `_bincount_kernel` | 92059.8 us | 1363.3 us |
| `_pack_prompt_bin_mask_kernel` | - | 1897.9 us |
| row zeroing | 335.0 us | 291.8 us |
| all measured work | 92444.9 us | 3601.3 us (**25.7x**) |

#### End-to-end, penalties enabled

Qwen3-0.6B on one A3, MRV2 with `VLLM_USE_V2_MODEL_RUNNER=1` and `--enforce-eager`;
32 requests, concurrency 8, 23230 prompt tokens, `max_tokens=16`. Requests use
`top_p=1.0, top_k=-1` to avoid an unrelated top-k/top-p Triton compile failure in the test
image. All 32 requests succeeded in every configuration.

| build | penalties | mean TTFT run 1 | mean TTFT run 2 |
|---|---|---:|---:|
| baseline | on | 1656.7 ms | 1689.9 ms |
| baseline | off | 1510.8 ms | 1569.4 ms |
| this PR | on | **1505.1 ms** | **1522.1 ms** |
| this PR | off | 1508.4 ms | 1519.1 ms |

The cost of enabling penalties changes from +145.9/+120.5 ms on baseline to
-3.3/+3.0 ms with this PR. Penalties-on TTFT improves by 9.2% and 9.9%; penalties-off is
the control for changes outside the optimized path.

#### CI and environments

- CI: pre-commit, mypy, coverage recommendation, test selection, main, and DCO pass.
  Selected NPU CI is pending a `ready-precise` or `ready-all` label.
- NPU: Atlas A2 910B1, CANN 9.0.0/9.1.0; Atlas A3 Ascend910_9392, CANN 9.0.1;
  triton_ascend 3.2.1; torch 2.10.0.
- GPU baseline: NVIDIA H20; torch 2.13.0+cu130; vLLM 0.27.1.

AI assistance was used during implementation and test drafting. The author reviewed every
changed line and ran the reported hardware tests.

- vLLM version: v0.27.1
- vLLM main: https://github.com/vllm-project/vllm/commit/ba07e4a48fc951300d97eb506217dd530583dea3

### CODE REVIEW

File: vllm_ascend/worker/v2/sample/penalties.py [hunk 1]
```diff
@@ -206,20 +234,41 @@ def bincount(
     output_bin_counts: torch.Tensor,
     max_prefill_len: int,
 ) -> None:
-    prompt_bin_mask[expanded_idx_mapping] = 0
     output_bin_counts[expanded_idx_mapping] = 0
     num_tokens = expanded_idx_mapping.shape[0]
+    num_words = prompt_bin_mask.shape[1]
```
[the remark this answers is not in the document]
Comment (zihanlin-ai) [reply vllm_ascend/worker/v2/sample/penalties.py:239]: Thanks for the suggestion. `bincount()` is called only when there are newly admitted penalty requests, so `expanded_idx_mapping` is non-empty on the production path. A zero-sized launch also predates this PR. Adding a new empty-input guard would broaden this performance-only change without being required for #14158, so I’m keeping the fix minimal.

File: vllm_ascend/worker/v2/sample/penalties.py [hunk 2]
```diff
@@ -206,20 +234,41 @@ def bincount(
...
+    vocab_size = output_bin_counts.shape[1]
     BLOCK_SIZE = 1024
     num_blocks = triton.cdiv(max_prefill_len, BLOCK_SIZE)
+    # Scratch flags for the prompt bitmask, one byte per vocabulary entry per
+    # request being counted. Freed on return; num_tokens is the number of newly
+    # added penalty requests, so this stays small.
+    prompt_flag = torch.zeros(
```
[the remark this answers is not in the document]
Comment (zihanlin-ai) [reply vllm_ascend/worker/v2/sample/penalties.py:246]: Thanks. This PR documents A2/A3 operator benchmarks, an A3 profiler comparison, and two E2E runs, while the committed nightly test covers deterministic correctness.

The repository does not currently provide a stable NPU performance-test harness or a cross-device latency threshold. Adding an absolute timing assertion here would be hardware- and toolchain-sensitive, while retaining the old atomic kernel solely for relative benchmarking would materially expand this focused change.

I am therefore keeping this fix minimal and leaving reusable performance-regression infrastructure to a separate change.
